File: push/src/main/java/com/xiaomi/smack/provider/ProviderManager.java
package com.xiaomi.smack.provider;

import com.xiaomi.smack.packet.IQ;
import com.xiaomi.smack.packet.PacketExtension;
import java.io.InputStream;
import java.net.URL;
import java.util.ArrayList;
import java.util.Collection;
import java.util.Collections;
import java.util.Enumeration;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;
import org.xmlpull.v1.XmlPullParser;
import org.xmlpull.v1.XmlPullParserFactory;

/* JADX INFO: loaded from: miuipushsdkshared_3_7_9.jar:com/xiaomi/smack/provider/ProviderManager.class */
public class ProviderManager {
    private static ProviderManager instance;
    private Map<String, Object> extensionProviders = new ConcurrentHashMap();
    private Map<String, Object> iqProviders = new ConcurrentHashMap();

    private ProviderManager() throws Throwable {
        initialize();
    }

    private ClassLoader[] getClassLoaders() {
        ClassLoader[] classLoaderArr = {ProviderManager.class.getClassLoader(), Thread.currentThread().getContextClassLoader()};
        ArrayList arrayList = new ArrayList();
        for (ClassLoader classLoader : classLoaderArr) {
            if (classLoader != null) {
                arrayList.add(classLoader);
            }
        }
        return (ClassLoader[]) arrayList.toArray(new ClassLoader[arrayList.size()]);
    }

    public static ProviderManager getInstance() {
        ProviderManager providerManager;
        synchronized (ProviderManager.class) {
            try {
                if (instance == null) {
                    instance = new ProviderManager();
                }
                providerManager = instance;
            } catch (Throwable th) {
                throw new RuntimeException(th);
            }
        }
        return providerManager;
    }

    private String getProviderKey(String str, String str2) {
        StringBuilder sb = new StringBuilder();
        sb.append("<");
        sb.append(str);
        sb.append("/>");
        if (str != null) {
            sb.append("<");
            sb.append(str2);
            sb.append("/>");
        }
        return sb.toString();
    }

    public static void setInstance(ProviderManager providerManager) {
        synchronized (ProviderManager.class) {
            try {
                if (instance != null) {
                    throw new IllegalStateException("ProviderManager singleton already set");
                }
                instance = providerManager;
            } catch (Throwable th) {
                throw th;
            }
        }
    }

    public void addExtensionProvider(String str, String str2, Object obj) {
        if (!(obj instanceof PacketExtensionProvider) && !(obj instanceof Class)) {
            throw new IllegalArgumentException("Provider must be a PacketExtensionProvider or a Class instance.");
        }
        this.extensionProviders.put(getProviderKey(str, str2), obj);
    }

    public void addIQProvider(String str, String str2, Object obj) {
        if (!(obj instanceof IQProvider) && (!(obj instanceof Class) || !IQ.class.isAssignableFrom((Class) obj))) {
            throw new IllegalArgumentException("Provider must be an IQProvider or a Class instance.");
        }
        this.iqProviders.put(getProviderKey(str, str2), obj);
    }

    public Object getExtensionProvider(String str, String str2) {
        return this.extensionProviders.get(getProviderKey(str, str2));
    }

    public Collection<Object> getExtensionProviders() {
        return Collections.unmodifiableCollection(this.extensionProviders.values());
    }

    public Object getIQProvider(String str, String str2) {
        return this.iqProviders.get(getProviderKey(str, str2));
    }

    public Collection<Object> getIQProviders() {
        return Collections.unmodifiableCollection(this.iqProviders.values());
    }

    protected void initialize() throws Throwable {
        try {
            for (ClassLoader classLoader : getClassLoaders()) {
                Enumeration<URL> resources = classLoader.getResources("META-INF/smack.providers");
                while (resources.hasMoreElements()) {
                    InputStream inputStream = null;
                    try {
                        InputStream inputStreamOpenStream = resources.nextElement().openStream();
                        XmlPullParser xmlPullParserNewPullParser = XmlPullParserFactory.newInstance().newPullParser();
                        xmlPullParserNewPullParser.setFeature("http://xmlpull.org/v1/doc/features.html#process-namespaces", true);
                        xmlPullParserNewPullParser.setInput(inputStreamOpenStream, "UTF-8");
                        int eventType = xmlPullParserNewPullParser.getEventType();
                        do {
                            if (eventType == 2) {
                                if (xmlPullParserNewPullParser.getName().equals("iqProvider")) {
                                    xmlPullParserNewPullParser.next();
                                    xmlPullParserNewPullParser.next();
                                    String strNextText = xmlPullParserNewPullParser.nextText();
                                    xmlPullParserNewPullParser.next();
                                    xmlPullParserNewPullParser.next();
                                    String strNextText2 = xmlPullParserNewPullParser.nextText();
                                    xmlPullParserNewPullParser.next();
                                    xmlPullParserNewPullParser.next();
                                    String strNextText3 = xmlPullParserNewPullParser.nextText();
                                    String providerKey = getProviderKey(strNextText, strNextText2);
                                    if (!this.iqProviders.containsKey(providerKey)) {
                                        try {
                                            Class<?> cls = Class.forName(strNextText3);
                                            if (IQProvider.class.isAssignableFrom(cls)) {
                                                try {
                                                    try {
                                                        this.iqProviders.put(providerKey, cls.newInstance());
                                                    } catch (Throwable th) {
                                                        th = th;
                                                        inputStream = inputStreamOpenStream;
                                                        try {
                                                            inputStream.close();
                                                        } catch (Exception e) {
                                                        }
                                                        throw th;
                                                    }
                                                } catch (ClassNotFoundException e2) {
                                                    e2.printStackTrace();
                                                }
                                            } else if (IQ.class.isAssignableFrom(cls)) {
                                                this.iqProviders.put(providerKey, cls);
                                            }
                                        } catch (ClassNotFoundException e3) {
                                            e3.printStackTrace();
                                        }
                                    }
                                } else if (xmlPullParserNewPullParser.getName().equals("extensionProvider")) {
                                    xmlPullParserNewPullParser.next();
                                    xmlPullParserNewPullParser.next();
                                    String strNextText4 = xmlPullParserNewPullParser.nextText();
                                    xmlPullParserNewPullParser.next();
                                    xmlPullParserNewPullParser.next();
                                    String strNextText5 = xmlPullParserNewPullParser.nextText();
                                    xmlPullParserNewPullParser.next();
                                    xmlPullParserNewPullParser.next();
                                    String strNextText6 = xmlPullParserNewPullParser.nextText();
                                    String providerKey2 = getProviderKey(strNextText4, strNextText5);
                                    if (!this.extensionProviders.containsKey(providerKey2)) {
                                        try {
                                            Class<?> cls2 = Class.forName(strNextText6);
                                            if (PacketExtensionProvider.class.isAssignableFrom(cls2)) {
                                                this.extensionProviders.put(providerKey2, cls2.newInstance());
                                            } else if (PacketExtension.class.isAssignableFrom(cls2)) {
                                                this.extensionProviders.put(providerKey2, cls2);
                                            }
                                        } catch (ClassNotFoundException e4) {
                                            e4.printStackTrace();
                                        }
                                    }
                                }
                            }
                            eventType = xmlPullParserNewPullParser.next();
                        } while (eventType != 1);
                        inputStreamOpenStream.close();
                    } catch (Throwable th2) {
                        throw th2;
                    }
                }
            }
        } catch (Exception e5) {
            e5.printStackTrace();
        }
    }

    public void removeExtensionProvider(String str, String str2) {
        this.extensionProviders.remove(getProviderKey(str, str2));
    }

    public void removeIQProvider(String str, String str2) {
        this.iqProviders.remove(getProviderKey(str, str2));
    }
}

File: legacy/src/main/java/com/xiaomi/smack/provider/ProviderManager.kt
package com.xiaomi.smack.provider

import com.xiaomi.smack.packet.IQ
import com.xiaomi.smack.packet.PacketExtension
import java.util.Collections
import java.util.concurrent.ConcurrentHashMap
import org.xmlpull.v1.XmlPullParser
import org.xmlpull.v1.XmlPullParserFactory

/*
 * Current override reference: com.xiaomi.xmsf 0.3.17-20260410000745 (versionCode 1003003000),
 * base.apk sha256 f3d72b6f5e1427ceecd3147a051d58e4dc95bb528397d486658e01cad9f7e590,
 * JADX path: com.xiaomi.xmsf/current/base/sources/com/xiaomi/smack/provider/ProviderManager.java
 * This runtime keeps local provider collection and singleton replacement helpers.
 */
class ProviderManager private constructor() {
    private val extensionProviders: MutableMap<String, Any> = ConcurrentHashMap()
    private val iqProviders: MutableMap<String, Any> = ConcurrentHashMap()

    init {
        initialize()
    }

    private fun getClassLoaders(): Array<ClassLoader> {
        val classLoaders = arrayOf(
            ProviderManager::class.java.classLoader,
            Thread.currentThread().contextClassLoader,
        )
        val result = ArrayList<ClassLoader>()
        for (classLoader in classLoaders) {
            if (classLoader != null) {
                result.add(classLoader)
            }
        }
        return result.toTypedArray()
    }

    private fun getProviderKey(str: String?, str2: String?): String {
        return buildString {
            append("<")
            append(str)
            append("/>")
            if (str != null) {
                append("<")
                append(str2)
                append("/>")
            }
        }
    }

    fun addExtensionProvider(str: String?, str2: String?, obj: Any) {
        if (obj !is PacketExtensionProvider && obj !is Class<*>) {
            throw IllegalArgumentException("Provider must be a PacketExtensionProvider or a Class instance.")
        }
        extensionProviders[getProviderKey(str, str2)] = obj
    }

    fun addIQProvider(str: String?, str2: String?, obj: Any) {
        if (obj !is IQProvider && (obj !is Class<*> || !IQ::class.java.isAssignableFrom(obj))) {
            throw IllegalArgumentException("Provider must be an IQProvider or a Class instance.")
        }
        iqProviders[getProviderKey(str, str2)] = obj
    }

    fun getExtensionProvider(str: String?, str2: String?): Any? = extensionProviders[getProviderKey(str, str2)]

    fun getExtensionProviders(): Collection<Any> = Collections.unmodifiableCollection(extensionProviders.values)

    fun getIQProvider(str: String?, str2: String?): Any? = iqProviders[getProviderKey(str, str2)]

    fun getIQProviders(): Collection<Any> = Collections.unmodifiableCollection(iqProviders.values)

    @Throws(Throwable::class)
    fun initialize() {
        try {
            for (classLoader in getClassLoaders()) {
                val resources = classLoader.getResources("META-INF/smack.providers")
                while (resources.hasMoreElements()) {
                    resources.nextElement().openStream().use { inputStream ->
                        val parser = XmlPullParserFactory.newInstance().newPullParser()
                        parser.setFeature("http://xmlpull.org/v1/doc/features.html#process-namespaces", true)
                        parser.setInput(inputStream, "UTF-8")
                        var eventType = parser.eventType
                        while (eventType != XmlPullParser.END_DOCUMENT) {
                            if (eventType == XmlPullParser.START_TAG) {
                                when (parser.name) {
                                    "iqProvider" -> loadIQProvider(parser)
                                    "extensionProvider" -> loadExtensionProvider(parser)
                                }
                            }
                            eventType = parser.next()
                        }
                    }
                }
            }
        } catch (e: Exception) {
            e.printStackTrace()
        }
    }

    private fun loadIQProvider(parser: XmlPullParser) {
        parser.next()
        parser.next()
        val elementName = parser.nextText()
        parser.next()
        parser.next()
        val namespace = parser.nextText()
        parser.next()
        parser.next()
        val className = parser.nextText()
        val providerKey = getProviderKey(elementName, namespace)
        if (iqProviders.containsKey(providerKey)) {
            return
        }
        try {
            val cls = Class.forName(className)
            if (IQProvider::class.java.isAssignableFrom(cls)) {
                iqProviders[providerKey] = cls.getDeclaredConstructor().newInstance()
            } else if (IQ::class.java.isAssignableFrom(cls)) {
                iqProviders[providerKey] = cls
            }
        } catch (e: ClassNotFoundException) {
            e.printStackTrace()
        }
    }

    private fun loadExtensionProvider(parser: XmlPullParser) {
        parser.next()
        parser.next()
        val elementName = parser.nextText()
        parser.next()
        parser.next()
        val namespace = parser.nextText()
        parser.next()
        parser.next()
        val className = parser.nextText()
        val providerKey = getProviderKey(elementName, namespace)
        if (extensionProviders.containsKey(providerKey)) {
            return
        }
        try {
            val cls = Class.forName(className)
            if (PacketExtensionProvider::class.java.isAssignableFrom(cls)) {
                extensionProviders[providerKey] = cls.getDeclaredConstructor().newInstance()
            } else if (PacketExtension::class.java.isAssignableFrom(cls)) {
                extensionProviders[providerKey] = cls
            }
        } catch (e: ClassNotFoundException) {
            e.printStackTrace()
        }
    }

    fun removeExtensionProvider(str: String?, str2: String?) {
        extensionProviders.remove(getProviderKey(str, str2))
    }

    fun removeIQProvider(str: String?, str2: String?) {
        iqProviders.remove(getProviderKey(str, str2))
    }

    companion object {
        private var instance: ProviderManager? = null

        @JvmStatic
        fun getInstance(): ProviderManager {
            synchronized(ProviderManager::class.java) {
                try {
                    if (instance == null) {
                        instance = ProviderManager()
                    }
                    return instance!!
                } catch (throwable: Throwable) {
                    throw RuntimeException(throwable)
                }
            }
        }

        @JvmStatic
        fun setInstance(providerManager: ProviderManager) {
            synchronized(ProviderManager::class.java) {
                if (instance != null) {
                    throw IllegalStateException("ProviderManager singleton already set")
                }
                instance = providerManager
            }
        }
    }
}
